src/esm.ts

```diff
@@ -147,12 +149,40 @@ export function createEsmHooks(tsNodeService: Service) {
       return defer();
     }
 
-    // pathname is the path to be resolved
+    let candidateSpecifiers: string[] = [specifier];
+
+    if (context.parentURL) {
+      const parentUrl = new URL(context.parentURL);
+      if (parentUrl.pathname && extname(parentUrl.pathname) === '.ts') {
+        const mappedSpecifiers = mapPath(specifier);
+        if (mappedSpecifiers) {
+          candidateSpecifiers = mappedSpecifiers.map((path) =>
+            pathToFileURL(path).toString()
+          );
+        }
+      }
+    }
+
+    let candidateSpecifier: string | undefined;
+    while ((candidateSpecifier = candidateSpecifiers.shift())) {
+      try {
+        return await nodeResolveImplementation.defaultResolve(
+          candidateSpecifier,
+          context,
+          defaultResolve
+        );
+      } catch (err) {
+        const isNotFoundError = (<any>err).code === 'ERR_MODULE_NOT_FOUND';
+        if (isNotFoundError && candidateSpecifiers.length > 0) {
+          continue;
+        } else {
+          throw err;
+        }
+      }
+    }
 
-    return nodeResolveImplementation.defaultResolve(
-      specifier,
-      context,
-      defaultResolve
+    throw new Error(
+      `Empty mapped paths for ${specifier} imported from ${context.parentURL}`
```

Comment:
Nit: add a code comment explaining it is a bug to ever get here.

Reply:
https://github.com/TypeStrong/ts-node/pull/1585/commits/2fcb9821593fd9235f45585ed0c95b1943a9c3de